### PR TITLE
planner: fix the panic that the index's range length may exceed its original column count

### LINKED ISSUE #25219: panic on DeriveStats for secondary index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `tab1` (
  `pk` int(11) NOT NULL,
  `col0` int(11) DEFAULT NULL,
  `col1` float DEFAULT NULL,
  `col2` text DEFAULT NULL,
  `col3` int(11) DEFAULT NULL,
  `col4` float DEFAULT NULL,
  `col5` text DEFAULT NULL,
  PRIMARY KEY (`pk`),
  KEY `idx_tab1_0` (`col0`),
  KEY `idx_tab1_1` (`col1`),
  KEY `idx_tab1_3` (`col3`),
  KEY `idx_tab1_4` (`col4`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
INSERT INTO tab1 (`pk`, `col0`, `col1`, `col2`, `col3`, `col4`, `col5`) VALUES
  (0, 5283, 6771.3e0, 'mkzsx', 1292, 8013.28e0, 'dxpup')
, (1, 3281, 3462.97e0, 'ulbku', 4126, 4213.14e0, 'zgiep')
, (2, 7435, 5423.44e0, 'bavre', 8205, 4321.4e0, 'pehqk')
, (3, 9743, 9466.82e0, 'nxqgw', 6827, 1253.25e0, 'ryfpv')
, (4, 5118, 3674.84e0, 'loabi', 6826, 7321.24e0, 'jpzeo')
, (5, 2626, 3649.98e0, 'zaogd', 9821, 480.79e0, 'wxeis')
, (6, 6751, 8763.58e0, 'pgfhq', 3466, 580.8e0, 'dkrzp')
, (7, 8632, 8407.99e0, 'wbgjr', 2700, 5617.88e0, 'kujdf')
, (8, 7319, 8241.65e0, 'wkodb', 998, 7978.63e0, 'wvsaw')
, (9, 1023, 6447.28e0, 'clpcj', 1620, 699.57e0, 'jnlfy')
;
ANALYZE TABLE tab1;
SELECT pk FROM tab1 WHERE (((col3 IS NULL OR (col1 IS NULL) AND (col3 <= 6659)))) AND ((col3 = 5815)) AND col4 = 6926.60;
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): runtime error: index out of range [1] with length 1

goroutine 1286 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x1053503f8, 0x14010fbe240, 0x14011026f00)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:799 +0x90
panic(0x1050f9f00, 0x140129c7ef0)
	/usr/local/go/src/runtime/panic.go:971 +0x424
github.com/pingcap/tidb/statistics.(*Index).expBackoffEstimation(0x14012ec7200, 0x14012f97200, 0x14012e37940, 0x14011b42900, 0x0, 0x14012e5c700, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/statistics/histogram.go:1296 +0xbb0
github.com/pingcap/tidb/statistics.(*Index).GetRowCount(0x14012ec7200, 0x14012f97200, 0x14012e37940, 0x140129957e0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/statistics/histogram.go:1256 +0x6dc
github.com/pingcap/tidb/statistics.(*HistColl).GetRowCountByIndexRanges(0x14012e37940, 0x14012f97200, 0x3, 0x140129957e0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/statistics/table.go:358 +0x348
github.com/pingcap/tidb/planner/core.(*DataSource).fillIndexPath(0x14012fb4280, 0x1401304de60, 0x14012e37900, 0x3, 0x4, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/logical_plans.go:852 +0x7ec
github.com/pingcap/tidb/planner/core.(*DataSource).DeriveStats(0x14012fb4280, 0x106bbcc10, 0x0, 0x0, 0x14012d20aa0, 0x106bbcc10, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/stats.go:272 +0x384
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).recursiveDeriveStats(0x14012fb42a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/stats.go:125 +0x368
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).recursiveDeriveStats(0x14012fc02c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/stats.go:118 +0x1a0
github.com/pingcap/tidb/planner/core.physicalOptimize(0x10537c860, 0x14012fc02a0, 0x14012e5c488, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:246 +0x58
github.com/pingcap/tidb/planner/core.DoOptimize(0x1053503f8, 0x14012e4d620, 0x10537d798, 0x14010b55180, 0x20a2, 0x10537c860, 0x14012fc02a0, 0x0, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:152 +0x264
github.com/pingcap/tidb/planner.optimize(0x1053503f8, 0x14012e4d620, 0x10537d798, 0x14010b55180, 0x105360a78, 0x14012ad6c30, 0x105377de8, 0x14011253360, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/optimize.go:288 +0xabc
github.com/pingcap/tidb/planner.Optimize(0x1053503f8, 0x14012e4d620, 0x10537d798, 0x14010b55180, 0x105360a78, 0x14012ad6c30, 0x105377de8, 0x14011253360, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/optimize.go:124 +0x70c
github.com/pingcap/tidb/executor.(*Compiler).Compile(0x14011152b18, 0x1053503f8, 0x14012e4d620, 0x1053654d8, 0x14012ad6c30, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/executor/compiler.go:61 +0x4b4
github.com/pingcap/tidb/session.(*session).ExecuteStmt(0x14010b55180, 0x1053503f8, 0x14012e4d620, 0x1053654d8, 0x14012ad6c30, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/session/session.go:1526 +0x664
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0x1401105fb60, 0x1053503f8, 0x14012e4d620, 0x1053654d8, 0x14012ad6c30, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/driver_tidb.go:218 +0x6c
github.com/pingcap/tidb/server.(*clientConn).handleStmt(0x14011026f00, 0x1053503f8, 0x14012e4d620, 0x1053654d8, 0x14012ad6c30, 0x106bbcc10, 0x0, 0x0, 0x1, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:1691 +0x184
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0x14011026f00, 0x105350350, 0x14012af9840, 0x14012b2d881, 0x78, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:1563 +0x800
github.com/pingcap/tidb/server.(*clientConn).dispatch(0x14011026f00, 0x105350350, 0x14012af9840, 0x14012b2d881, 0x78, 0x78, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/
```

### 4. What is your TiDB version? (Required)

47f0f15b14ed54fc2222f3e304e29df7b05e6805

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #25219  

Problem Summary:

Since we are encoding the primary key as the key part in its key-value store for some index, the index's range would exceed its original column count.
So we need to adjust the things we maintain.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

Add the pk to the end if needed.

How it Works:

Then we can use the pk to estimate and avoid the panic.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: planner/core/logical_plans.go [hunk 1]
```diff
@@ -835,6 +835,10 @@ func (ds *DataSource) fillIndexPath(path *util.AccessPath, conds []expression.Ex
 			if !alreadyHandle {
 				path.IdxCols = append(path.IdxCols, handleCol)
 				path.IdxColLens = append(path.IdxColLens, types.UnspecifiedLength)
+				// Also updates the map that maps the index id to its prefix column ids.
+				if len(ds.tableStats.HistColl.Idx2ColumnIDs[path.Index.ID]) == len(path.Index.Columns) {
+					ds.tableStats.HistColl.Idx2ColumnIDs[path.Index.ID] = append(ds.tableStats.HistColl.Idx2ColumnIDs[path.Index.ID], handleCol.UniqueID)
```
Comment (time-and-fate) [inline planner/core/logical_plans.go:840]: I think this would affect many other uses of `Idx2ColumnIDs`, like `getGroupNDVs()`.

**Review by time-and-fate — COMMENTED** [review]
And there are fails and panics in the tests.

File: planner/core/logical_plans.go (hunk 1 above)
Comment (winoros) [reply planner/core/logical_plans.go:840]: The GroupNDV checks that `len(Idx2ColumnIDs[idxID]) == len(idxInfo.Columns)`

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**